_keystone/oidc-federation.md

```diff
@@ -0,0 +1,503 @@
+empty
+any_ony_of
+not_any_of
+blacklist
+whitelist
+```
+
+To set the username we can use **{0}** in the local section to indicate to use a field from the remote section. Then we can additional information to the new federated username. We can reference the remote section in the local section by index ("{i}", where i is the index of the remote field):
+
+```
+[
+  {
+    "local": [
+      {
+        "user": {
+          "name": "CompanyXYZ/{0}"
+        }
+      },
+               ...
+    ],
+    "remote": [
+      {
+        "type": "HTTP_OIDC_CLAIM_EMAIL"
+      },
+               ...
+    ]
+  }
+]
+```
+
+
+
+Here are some example scenarios:
+
+**Example 1.**
+Here we are mapping users from the CompanyXYZ IDP and allowing any user
+from the `admin_group`. The user's email address will be used to create
+a user on Keystone. If the user's email is `test@test.ibmcloud.com`, then
+the resulting username would be `CompanyXYZ/test@test.ibmcloud.com`.
+
+```
+[
+  {
+    "local": [
+      {
+        "user": {
+          "name": "CompanyXYZ/{0}"
+        }
+      },
+      {
+        "group": {
+          "domain": {
+            "name": "Default"
+          },
+          "name": "company_xyz_admin_group"
+        }
+      }
+    ],
+    "remote": [
+      {
+        "type": "HTTP_OIDC_CLAIM_EMAIL"
+      },
+      {
+        "type": "HTTP_OIDC_CLAIM_GROUP",
+        "any_one_of": [
+          "admin_group"
+        ]
+      }
+    ]
+  }
+]
+```
+
+**Example 2.** Multiple Groups
+
+In the following example we are using
+
+```
+[
+  {
+    "local": [
+      {
+        "user": {
+          "name": "CompanyXYZ/{0}"
+        }
+      },
+      {
+        "group": {
+          "domain": {
+            "name": "Default"
+          },
+          "name": "company_xyz_admin_group"
+        }
+      }
+    ],
+    "remote": [
+      {
+        "type": "HTTP_OIDC_CLAIM_EMAIL"
+      },
+      {
+        "type": "HTTP_OIDC_CLAIM_GROUP",
+        "any_one_of": [
+          "admin_group"
+        ]
+      }
+    ]
+  },
+ {
+    "local": [
+      {
+        "user": {
+          "name": "CompanyXYZ/{0}"
+        }
+      },
+      {
+        "group": {
+          "domain": {
+            "name": "Default"
+          },
+          "name": "company_xyz_member_group"
+        }
+      }
+    ],
+    "remote": [
+      {
+        "type": "HTTP_OIDC_CLAIM_EMAIL"
+      },
+      {
+        "type": "HTTP_OIDC_CLAIM_GROUP",
+        "any_one_of": [
+          "member_group"
+        ]
+      }
+    ]
+  }
+]
+```
+
+
+## <a name="using_horizon"></a>Using Horizon
+ The OpenStack dashboard on the Identity Provider should have a drop-down menu labeled **Authenticate with SSO**.
+ 1. The dropdown should show a label for the IdP. Select the IDP and click Connect.
+ 2. You will then be redirected to the IDP log in page.
+ 3. After successfully authenticating with the IDP, you should then redirected to the horizon page.
+
+## <a name="using_oauth_cli"></a>Using OAuth CLI
```

Comment:
Title says CLI but this is showing how to use the API. We should probably look at that as well.